File: ProblemSets/LeetCode/138.cc
#include <bits/stdc++.h>
// Definition for a Node.
#ifdef NATIVE_IDE
class Node {
   public:
    int val;
    Node* next;
    Node* random;

    Node( int _val ) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
#endif
using namespace std;
class Solution {
   public:
    Node* copyRandomList( Node* head ) {
        vector<int> valList;
        unordered_map<Node*, int> Addr2Id;
        vector<Node*> oldAddrList;
        Node* p = head;
        while ( p ) {
            Addr2Id[p] = valList.size();
            oldAddrList.push_back( p );
            valList.push_back( p->val );
            p = p->next;
        }
        Node* resNode = NULL;
        if ( head )
            resNode = new Node( head->val );
        p = resNode;
        vector<Node*> newLinkedList{ resNode };
        for ( int i = 1; i < valList.size(); i++ ) {
            p->next = new Node( valList[i] );
            newLinkedList.push_back( p->next );
            p = p->next;
        }
        for ( int i = 0; i < valList.size(); i++ ) {
            if ( oldAddrList[i]->random )
                newLinkedList[i]->random = newLinkedList[Addr2Id[oldAddrList[i]->random]];
        }
        return resNode;
    }
};

File: ProblemSets/LeetCode/Java/138.java
package LeetCode.Java;
// Definition for a Node.

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {

    public Node copyRandomList(Node head) {
        List<Integer> valList = new ArrayList<>();
        HashMap<Node, Integer> Addr2Id = new HashMap<>();
        List<Node> oldAddrList = new ArrayList<>();
        Node p = head;
        while (p != null) {
            Addr2Id.put(p, valList.size());
            oldAddrList.add(p);
            valList.add(p.val);
            p = p.next;
        }
        Node resNode = null;
        if (head != null)
            resNode = new Node(head.val);
        p = resNode;
        List<Node> newLinkedList = new ArrayList<>(); // List.of can't accept mull cause if can't deduce element type.
        newLinkedList.add(p);
        for (int i = 1; i < valList.size(); i++) {
            p.next = new Node(valList.get(i));
            newLinkedList.add(p.next);
            p = p.next;
        }
        for (int i = 0; i < valList.size(); i++) {
            if (oldAddrList.get(i).random != null)
                newLinkedList.get(i).random = newLinkedList.get(Addr2Id.get(oldAddrList.get(i).random));
        }
        return resNode;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        s.copyRandomList(new Node(10));
    }
}